### PR TITLE
Add syllabus classifier job

### PR DESCRIPTION
**Pull request by jodizzle**
- Bumps and adds various requirements to `requirements.txt`.
- Adds an `osp_pipeline/classifier` module.
    - Adds a "DocumentSyllabus" model with `document_id` and `is_syllabus` fields.
    - Adds a join, adapted from the one in [`osp_pipeline/institutions`](https://github.com/opensyllabus/osp-pipeline/blob/6bd66f4c14af43e65bee5b1da2659428c1db5256/osp_pipeline/institutions/joins.py).  The join gathers the text and metadata necessary for making predictions.
    - Adds a job to classify documents as either syllabus or not syllabus, storing the boolean value in the `is_syllabus` field.
- Installs `document_syllabus` into the config system (I think).
- Adds a rudimentary test for the job.

As best I can tell, this job functions as intended, so I'm opening it up for review.  We can run it as needed, but we should probably hold off on merging it until more thorough tests are added, and the approach behind loading the classifier pickle (`SyllabusClassfier.load()`) is re-tooled so that the pickle of the classifier is more closely tied to the version of `osp-pipeline` or to the Docker deployment.

Any comments about the general approach taken by the job would be appreciated.  (It's kind of akward -- there are a lot of transformations to wrangle the dataframe into a shape that the classifier pipeline likes.  I left a comment string in there about it.)

### CODE REVIEW

File: osp_pipeline/classifier/jobs/find_syllabi.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    # to the spark job causes errors (stuff about thread locking and
+    # being unable to pickle).
+    pipeline = classifier.pipeline
+
+    docs = Document.read(spark)
+    texts = DocumentText.read(spark)
+
+    match_docs = MatchDocument.build_df(docs, texts)
+
+    """
```
Comment (wearpants) [inline osp_pipeline/classifier/jobs/find_syllabi.py]: Why is this a bare string instead of a block comment?

File: osp_pipeline/classifier/jobs/find_syllabi.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    columns = MatchDocument.fields
+    new_names = {
+        'url': "WARC-Target-URI",
+        'source_url': "X-Source-Url",
+        'source_anchor': "X-Source-Anchor",
+        'text': "Text"
+    }
+
+    # XXX: Efficiency concerns?
```
Comment (wearpants) [inline osp_pipeline/classifier/jobs/find_syllabi.py]: This can be written as pure dataframe function instead of rdd + Python, I can help with that

File: osp_pipeline/classifier/jobs/find_syllabi.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    result = (
+        match_docs
+        .select(*[
+            wrapInArray(column).alias(new_names[column]) if column != 'id'
```
Comment (wearpants) [inline osp_pipeline/classifier/jobs/find_syllabi.py]: Err, this part at least

File: osp_pipeline/classifier/jobs/find_syllabi.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            else column
+            for column in columns
+        ])
+        .rdd.map(lambda row: classify_syllabi(row, pipeline))
```
Comment (wearpants) [inline osp_pipeline/classifier/jobs/find_syllabi.py]: This would be a UDF

File: osp_pipeline/classifier/jobs/find_syllabi.py (hunk 1 above)
Comment (jodizzle) [reply osp_pipeline/classifier/jobs/find_syllabi.py]: No particular reason, I just found it easier to format that way.  I'll certainly change it to block comments/remove it/make it a docstring before merging.

File: osp_pipeline/classifier/jobs/find_syllabi.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        'url': "WARC-Target-URI",
+        'source_url': "X-Source-Url",
+        'source_anchor': "X-Source-Anchor",
+        'text': "Text"
+    }
+
+    # XXX: Efficiency concerns?
+    result = (
+        match_docs
+        .select(match_docs.id, *[
```
Comment (wearpants) [inline osp_pipeline/classifier/jobs/find_syllabi.py]: I would make a little lambda to generate the when/otherwise/alias rename clause object, and call that in the *args instead of defining it inline - just for clarity purposes, it's a little hard to parse what's going on with so much nesting.

File: osp_pipeline/classifier/jobs/find_syllabi.py (hunk 2 above)
Comment (wearpants) [reply osp_pipeline/classifier/jobs/find_syllabi.py]: this updated broadcast + udf looks great!

File: osp_pipeline/classifier/jobs/find_syllabi.py (hunk 5 above)
Comment (jodizzle) [reply osp_pipeline/classifier/jobs/find_syllabi.py]: I moved the conditional stuff into a separate function.  Currently it's a pretty direct translation; I could also re-tool the parameters it takes, move it to a different part of the file, etc.

Personally, I'd be fine leaving it inline, but I'm amenable.

**Comment by wearpants** [comment]
Looks good

**Comment by jodizzle** [comment]
I rebased this PR and updated it to include more thorough testing of the syllabus classifier job.  I was considering putting the test revamp in a separate PR, but I felt it was easier and more appropriate to put it here.

There's a little bit of stuff that could be tweaked/cleaned up before merging, but I think the bulk of everything is in place.

- Adds some helper `definitions.py` files to provide absolute paths to parts of the directory tree (in this case, fixtures and classifier pickles, though this strategy could be applied more broadly).  This approach is something that could probably be cleaned up.
- Tests the classifier against all the fixtures in the document fixtures submodule.  In particular, feed the `find_syllabi` job some pre-made `Document` and `DocumentText` parquet tables, and then check the result against the entries of a `fixtures_cases.yml` file.  Each entry contains a True or False value for `is_syllabus`.  (Note: I made sure to check that these `is_syllabus` values were the output of both the `find_syllabi` job and the "raw" classifier outside of the job, since the two should definitely match up.)
- Lock the version of osp-pipeline to a version of the syllabus classifier by introducing a submodule to contain pickles of classifiers.  The syllabus classifier will be loaded exclusively from the `osp_pipeline/classifier/pickles` folder.

File: osp_pipeline/classifier/jobs/find_syllabi.py (hunk 1 above)
Comment (wearpants) [reply osp_pipeline/classifier/jobs/find_syllabi.py]: please make this a comment, not a string

File: osp_pipeline/classifier/definitions.py [hunk 6]
```diff
@@ -0,0 +1,4 @@
+import os
```
Comment (wearpants) [inline osp_pipeline/classifier/definitions.py]: I'd put the entire contents in `__init__.py`, no need for a separate file here

File: osp_pipeline/classifier/models.py [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+from osp_classifier.classifiers import SyllabusClassifier
+
+from osp_pipeline.core.models import Model
+
+
+class DocumentSyllabus(Model):
+
+    s3_key = 'document_syllabus'
+
+    schema = T.StructType([
```
Comment (wearpants) [inline osp_pipeline/classifier/models.py]: I might just do this as the table of document_ids that are syllabi, rather than all ids+boolean `is_syllabus`; that will be the most common use, and faster than a full join + filter

File: tests/definitions.py [hunk 8]
```diff
@@ -0,0 +1,4 @@
+import os
```
Comment (wearpants) [inline tests/definitions.py]: similiarly, just put the FIXTURES_ROOT directly in test_find_syllabi.py

File: osp_pipeline/classifier/models.py (hunk 7 above)
Comment (jodizzle) [reply osp_pipeline/classifier/models.py]: Hmm, this is an interesting idea, I'll think on it for a little bit.  It would require some restructuring of the job and tests, though maybe for the best.

**Comment by jodizzle** [comment]
@wearpants Thanks for the note about `.toJSON` -- I had tried to write to JSON before with `spark.write.json` for the benefits you identified, but it was outputting across multiple JSONs in an annoying way.  Using `.toJSON` and some tricks allows me to put everything into one big JSON for each model.

However, there's a potential annoyance here: like every job, `find_syllabi(sc, spark)` reads the dumps on it's own (hence the need to patch the config results paths), and that in turn is [glued to reading from parquet tables](https://github.com/opensyllabus/osp-pipeline/blob/15f840e42c7afb98d48c96b74b252dd820e1d67c/osp_pipeline/core/models.py#L79).  I could read from JSON dumps, then re-dump to parquet tables, and then `find_syllabi` would read those, but I think that's pretty inefficient?

Would it be possible/feasible to have `Model.read` read from a generic source?

**Comment by wearpants** [comment]
Would it be possible/feasible to have Model.read read from a generic source?


Let's do this functionality in `load_df` instead: add keyword only args
`path` (default None) and `format` (default "parquet"). If path is None,
fall back to the path from config and issue a warning (with warnings module)

**Comment by jodizzle** [comment]
So I guess that means we should add extra parameters to `find_syllabi` as well?

> fall back to the path from config and issue a warning (with warnings module)

What is the warnings module?  Or do you just mean `log.warning`?

**Comment by wearpants** [comment]
https://docs.python.org/3/library/warnings.html - you'll want to set the
stacklevel to refer to the calller of load_df(path=None), so we can
eventually fix those places.

File: osp_pipeline/classifier/jobs/find_syllabi.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+from osp_pipeline.classifier.models import DocumentSyllabus
+from osp_pipeline.classifier.joins import MatchDocument
+from osp_pipeline.corpus.models import Document, DocumentText
+
+
+def find_syllabi(docs, texts):
+    """Create a dataframe of information from a syllabus classifier.
+
+    Args:
+        docs (dataframe): A Document dataframe.
```
Comment (wearpants) [inline osp_pipeline/classifier/jobs/find_syllabi.py]: please use sphinx style docstrings

File: osp_pipeline/classifier/jobs/find_syllabi.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+            .alias(new_names[column])
+        )
+
+    result = (
+        match_docs
+        .select(match_docs.id, *[
+            wrap_and_rename(column)
+            for column in to_rename
+        ])
+        .withColumn(
```
Comment (wearpants) [inline osp_pipeline/classifier/jobs/find_syllabi.py]: to make the change to a table of syllabus_id's only containing known syllabi, change this from adding a new column to `filter(..)`ing on predictWithPipeline

File: .gitmodules [hunk 11]
```diff
@@ -4,3 +4,6 @@
 [submodule "deploy/roles/docker"]
 	path = deploy/roles/docker
 	url = https://github.com/angstwad/docker.ubuntu.git
+[submodule "osp_pipeline/classifier/pickles"]
```
[the remark this answers is not in the document]
Comment (jodizzle) [reply .gitmodules]: Ah, those are different repositories.  The one in requirements, `osp-classifier`, has some very simple classes that just act as wrappers around sklearn pipelines.  This submodule, `osp-pipeline-classifiers`, contains the pickles used in the pipeline.

If you think we could have a better name for the submodule than `osp-pipeline-classifiers`, let me know.

**Comment by wearpants** [comment]
I'd call it osp-pipeline-classifier-blobs... where is the version pinned
though?

I still think we should just include these directly in pipeline repo, do we
use them anywhere else? Don't understand why they're separate, let's
discuss on call

On Jan 22, 2018 2:57 PM, "jodizzle" <notifications@github.com> wrote:

*@jodizzle* commented on this pull request.
------------------------------

In .gitmodules
<https://github.com/opensyllabus/osp-pipeline/pull/105#discussion_r163051614>
:

> @@ -4,3 +4,6 @@
 [submodule "deploy/roles/docker"]
 	path = deploy/roles/docker
 	url = https://github.com/angstwad/docker.ubuntu.git
+[submodule "osp_pipeline/classifier/pickles"]

Ah, those are different repositories. The one in requirements,
osp-classifier, has some very simple classes that just act as wrappers
around sklearn pipelines. This submodule, osp-pipeline-classifiers,
contains the pickles used in the pipeline.

If you think we could have a better name for the submodule than
osp-pipeline-classifiers, let me know.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/opensyllabus/osp-pipeline/pull/105#discussion_r163051614>,
or mute the thread

**Comment by wearpants** [comment]
Hmm, unless the idea is the blobs are  private repo...

Ok, let's call it osp-pipeline-assets (one of these is more than enough)

Same question about pinning

**Comment by jodizzle** [comment]
@wearpants

> Hmm, unless the idea is the blobs are private repo...

Yes, the idea is to use a private repo.  In addition, if at some point we have a different solution for loading classifiers, we don't have to lug around large binary snapshots in the main git history.

> Ok, let's call it osp-pipeline-assets (one of these is more than enough)

Sure, that sounds reasonable (though I'm not sure what other assets we'd be adding).

> Same question about pinning

The pinning is handled just by how submodules work.  You add an entry to the `.gitmodules` folder, and then you `git add` specific commit hashes of the submodule if and when you need to update.  The reference to the submodule is then glued to that commit.  For example, [here's me updating the fixtures submodule to contain JSON dumps](https://github.com/opensyllabus/osp-pipeline/pull/105/commits/56aa207f14acfdb7cf0b3e2b96085490168e3ec3#diff-da0b2b0b955ea5f4cd5a9eeaadb8cb79), and [here's me adding the submodule with the classifier pickles](https://github.com/opensyllabus/osp-pipeline/pull/105/commits/2f4da9ee642f15ed9777ee9541d045ca59a8cb2b#diff-6857dccfe5de12c909c2fabe8d86aa0d).  We can play around with the contents of the submodules in local testing, but the references to them won't update unless we expressly tell them to.

File: osp_pipeline/classifier/jobs/find_syllabi.py (hunk 10 above)
Comment (jodizzle) [reply osp_pipeline/classifier/jobs/find_syllabi.py]: I used the `.where` alias instead, but yeah, same idea.

File: osp_pipeline/classifier/jobs/find_syllabi.py (hunk 9 above)
Comment (jodizzle) [reply osp_pipeline/classifier/jobs/find_syllabi.py]: Sure.  For reference, I think Sphinx supports the style of docstring I was using here (with an extension).

**Comment by jodizzle** [comment]
I believe I've addressed everything.  I think switching the `DocumentSyllabus` schema to be just a list of IDs does indeed simplify things.

There are two main things that I think should still be done with the actual job:

- There's a small preprocessing step of removing stubs (documents that don't have very much extracted text) that's done when training the classifier, and which should be used (in most contexts) when predicting with the classifier.
- By default, the classifier makes predictions based on an 0.5 threshold value: anything above that is marked as a syllabus, and anything below is marked as not a syllabus.  However, it is sometimes valuable to predict with a different threshold value, like when trying to control the number of false negatives/positives.  With Pandas dataframes, that can be done with e.g., `pipeline.predict_proba(X)[:, 1] > threshold`.

I'll work on the best ways to accomplish these in `find_syllabi`.

**Comment by wearpants** [comment]
Might it be useful to instead save the predicted value along with
document-id, so that we could play with different thresholds without
needing to recalc scores (by doing the filter step when we use the
classified syllabi, instead of saving only a Y/N) Would that make sense/be
useful? Realize this may shift documentsyllabus schema again....

On Jan 27, 2018 4:56 AM, "jodizzle" <notifications@github.com> wrote:

I believe I've addressed everything. I think switching the DocumentSyllabus
schema to be just a list of IDs does indeed simplify things.

There are two main things that I think should still be done with the actual
job:

   - There's a small preprocessing step of removing stubs (documents that
   don't have very much extracted text) that's done when training the
   classifier, and which should be used (in most contexts) when predicting
   with the classifier.
   - By default, the classifier makes predictions based on an 0.5 threshold
   value: anything above that is marked as a syllabus, and anything below is
   marked as not a syllabus. However, it is sometimes valuable to predict with
   a different threshold value, like when trying to control the number of
   false negatives/positives. With Pandas dataframes, that can be done with
   e.g., pipeline.predict_proba(X)[:, 1] > threshold.

I'll work on the best ways to accomplish these in find_syllabi.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/opensyllabus/osp-pipeline/pull/105#issuecomment-360974017>,
or mute the thread

**Comment by jodizzle** [comment]
Well, it's sort of a tradeoff.  If we switch to storing the result of `pipeline.predict_proba`, we lose the simplicity and efficiency of having a dataframe that's just a list of `document_id`s of known syllabi.  However, we gain access to a lot more information about our documents.

Having thought about it, I think storing the numeric prediction value is better.  As you suggest, it should make testing different thresholds easier, and it could more generally help with getting stats about the classifier's performance on our corpus.  Plus, it would make the unit tests on document fixtures more precise.  In terms of other parts of the pipeline that will use the dataframe, maybe we could just have a function somewhere in the `classifier` module that loads and returns a trimmed dataframe according to a `threshold` parameter?

Moving back to a two-column schema shouldn't be that much work.

**Comment by wearpants** [comment]
yeah, let's store the prediction values, and write a little function like
you describe

**Comment by jodizzle** [comment]
I've made some more updates:

- Switched the schema to be based on storing the raw probabilities output from the classifier, as discussed.
- Added `try_or_log` protection to the prediction step.  Errors will be filtered out, as usual.
- Included a small preprocessing step that mirrors the one the classifier uses before training, as discussed.
- Added a helper method, `DocumentSyllabus.with_threshold`, that will filter a DocumentSyllabus dataframe (or anything with a superset of the DocumentSyllabus schema, I suppose) by the `probability` field.
- Updated the structure and content of the tests to accommodate the above changes.

I think this PR is in a pretty good state, so I'll remove the "DON'T MERGE" stuff.  There are certain aspects to the job and model that probably should be generalized/made more flexible (I've left plenty of notes indicating this), but those can be addressed in future updates, and probably until after we've run the job.

File: osp_pipeline/classifier/models.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+        T.StructField('document_id', T.StringType(), nullable=False),
+        T.StructField('probability', T.DoubleType()),
+    ])
+
+    @classmethod
+    def with_threshold(cls, threshold=0.5, df=None):
```
Comment (wearpants) [inline osp_pipeline/classifier/models.py]: can we call this `above_threshold`

File: osp_pipeline/classifier/models.py [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+        """Return a DocumentSyllabus dataframe with probabilities > threshold.
+
+        Using this method with a threshold value of 0.0 or 0.5 (the
```
Comment (wearpants) [inline osp_pipeline/classifier/models.py]: this doesn't make sense, is it 0.0 or 0.5?

File: osp_pipeline/classifier/models.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+        gives higher recall, and raising the threshold gives higher
+        precision.
+
+        :param threshold: A float between 0.0 and 1.0.  Default: 0.5
+        :param df: The dataframe to filter.  If `None`, the dataframe is
+            read from the DocumentSyllabus config path.  Default:
+            `None`.
+        :returns: A DataFrame filtered by threshold.
+        """
+        if df is None:
```
Comment (wearpants) [inline osp_pipeline/classifier/models.py]: let's make it so you always have to pass `df` instead of loading implicitly.

File: osp_pipeline/classifier/models.py [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+    """
+
+    s3_key = 'document_syllabus'
+
+    schema = T.StructType([
+        T.StructField('document_id', T.StringType(), nullable=False),
+        T.StructField('probability', T.DoubleType()),
+    ])
+
+    @classmethod
```
Comment (wearpants) [inline osp_pipeline/classifier/models.py]: this can be a `staticmethod`, `cls` isn't used anywherre

File: osp_pipeline/classifier/jobs/find_syllabi.py [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+    characters are removed from consideration.
+
+    NOTE: Estimation is done with the `predict_proba` method, which
+    means the pipeline must support this method.
+
+    :param pipeline: An sklearn pipeline.
+    :param row: A pyspark.sql.Row.
+    :returns: A float in [0.0, 1.0].
+    """
+    if len(row['Text'][0]) <= 150:
```
Comment (wearpants) [inline osp_pipeline/classifier/jobs/find_syllabi.py]: let's make this `150` a module-level constant instead of hardcoded down here, call it `MIN_TEXT_LENGTH`

File: osp_pipeline/classifier/models.py (hunk 15 above)
Comment (jodizzle) [reply osp_pipeline/classifier/models.py]: Wups, missed this.

File: osp_pipeline/classifier/models.py (hunk 13 above)
Comment (jodizzle) [reply osp_pipeline/classifier/models.py]: Sorry, this was a silly detail of a different implementation that I accidentally left in the comment here.  It should be 0.5.

File: osp_pipeline/classifier/models.py (hunk 12 above)
Comment (jodizzle) [reply osp_pipeline/classifier/models.py]: Sure, as long as you don't think it's too wordy.

File: osp_pipeline/classifier/models.py (hunk 14 above)
Comment (jodizzle) [reply osp_pipeline/classifier/models.py]: Sounds good.  I was a bit unsure of the whether to include this clause, but I guess the safe approach is to just always pass it.

**Comment by jodizzle** [comment]
@wearpants I think I addressed all the comments.

File: osp_pipeline/classifier/jobs/find_syllabi.py [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+    NOTE: Estimation is done with the `predict_proba` method, which
+    means the pipeline must support this method.
+
+    :param pipeline: An sklearn pipeline.
+    :param row: A pyspark.sql.Row.
+    :returns: A float in [0.0, 1.0].
+    """
+    if len(row['Text'][0]) <= MIN_TEXT_LENGTH:
+        log.info((f"Filtered out 'document_id={doc_id}', extracted text <= "
+                   "150 characters."))
```
Comment (wearpants) [inline osp_pipeline/classifier/jobs/find_syllabi.py]: this should also use the constant...

File: osp_pipeline/classifier/models.py [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+
+    s3_key = 'document_syllabus'
+
+    schema = T.StructType([
+        T.StructField('document_id', T.StringType(), nullable=False),
+        T.StructField('probability', T.DoubleType()),
+    ])
+
+    @staticmethod
+    def with_threshold(df, above_threshold=0.5):
```
Comment (wearpants) [inline osp_pipeline/classifier/models.py]: no, the method name should be `above_threshold`, the argument can be `threshold`

**Comment by jodizzle** [comment]
Sorry, I was moving a bit too fast.  Should be fixed up now.
